Track busy/idle state for integration tests

jesseduffield opened the pull request:
For a better explanation than this now-outdated PR description gives, see [here](https://github.com/jesseduffield/lazygit/blob/track-pending-actions/docs/dev/Busy.md?rgh-link-date=2023-07-08T12%3A40%3A30Z)

This PR tracks pending tasks so that integration tests only run assertions and send keypresses when lazygit isn't in the middle of something.

We could use this logic in-app to create a generic loader but I like the fact that all our loaders currently have labels telling you what's happening, and I'm happy to just have it be a pattern that we should always use loaders for long-running actions.

The logic works like this: we maintain a global counter that tells us whether there's a pending action. If we're on the UI thread the counter will always be above zero, and if we're running something in the background (this is done via `OnBackgroundThread()`) we increment the counter before spawning a goroutine and then within that goroutine we defer decrementing the counter again. This means that the counter should be zero if nothing is happening (except for long-running goroutines that aren't associated with actions).

The following stuff happens on the UI thread:
* handling of keypresses
* handling of terminal events like resizes
* functions passed to the OnUIThread() function

Background stuff is done via `OnBackgroundThread`

If you want to spawn a goroutine that is long-running and doesn't relate to actions you can just spawn a regular goroutine.

There are a couple places where we need to manually increment and decrement the action counter, because goroutines are involved that stick around despite us no longer considering an action to be pending:
* When rendering to the main view, we increment the counter before we run the command and we decrement after we've loaded the first page, given that we don't bother loading anything more until the user themselves scrolls.
* When running a git command that can request credentials, we decrement the counter before prompting the user for a credential, and then increment it again afterwards. This means we're not telling the user that we're loading something when we're actually waiting on them to enter a credential. It's assumed (reasonably) that the counter will start above zero given that the git command is running in the first place.

If the counter ever goes below zero, there's a bug somewhere, and so we panic. This may be overkill but I really don't want any bugs in this code.

At the moment I'm seeing if we can remove all exponential backoffs in assertions by having tests progress whenever we're told that there are no more pending actions. This will speed things up and prevent false positives. But so far it's a bit flakey.

Things I'm not sure about:
* Using the 'Thread' terminology is a bit weird given we're dealing with goroutines, not threads, but 'UI Thread' is something everybody understands
* It's probably possible to refactor the two places we're manually incrementing/decrementing the counter to use the existing helper method 'OnBackgroundThread' but when I gave this a go in the context of git commands with credentials, it got pretty complicated pretty quickly.



- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go run scripts/cheatsheet/main.go generate`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [x] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [x] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

stefanhaller commented:
This is great!

There seems to be a problem at startup, somehow. I ran all tests in a loop to see if I get any to fail occasionally, and the ones that I have seen fail so far are `initial_open.go` and `push_follow_tags.go`. In both cases, they failed on the first check in their `Run` function, and in both cases I got an empty screen as output of `Final Lazygit state:`.

jesseduffield commented:
@stefanhaller thanks :)

My current theory with the failing tests is that there is a race condition where before we first load everything and render the screen, there is some other event that happens on the UI thread or a background thread which begins and completes before anything else happens, meaning that the test is notified that lazygit isn't busy and so it runs the first assertion before everything is ready.

I've spent a bit of time looking for the cause of the issue with no luck so far (I can't actually get failures locally). But I'll keep looking!

stefanhaller commented:
> I can't actually get failures locally

Interesting. Doing `while go test pkg/integration/clients/*.go; do : ; done` always fails for me eventually; sometimes after three iterations, sometimes after like 12. Never had to wait more than 20 iterations so far.

jesseduffield commented:
@stefanhaller ah nice I'll give that a go

jesseduffield commented:
I've fixed those bugs, cleaned up my commits, and added a [doc](https://github.com/jesseduffield/lazygit/blob/track-pending-actions/docs/dev/Busy.md) explaining the logic.

Some good news to report: tests are reliably passing even if I disable test retries! I've also fixed one of the issues causing an index.lock error, and I've re-enabled a test which was disabled due to that error.

jesseduffield commented:
(Scratch that: the index.lock test still fails locally. But other tests with intermittent index.lock failures were indeed fixed!)

jesseduffield commented:
@stefanhaller any thoughts on the code before I merge?

jesseduffield commented:
I've actually just pushed a new commit which changes the approach, going from a global counter to a global map of 'tasks' which can each be paused/continued. This is much more discoverable, but at the cost of bloating some function signatures.

stefanhaller commented on docs/dev/Busy.md:
```diff
@@ -0,0 +1,88 @@
+# Knowing when Lazygit is busy/idle
```
This is fantastic, thanks for taking the time to write this up. Unfortunately, a lot of this is now out of date after the last commit you pushed; it would be great to adapt this document to the current state.

stefanhaller commented:
When running the tests in a loop again with the latest version (f6d49b53d3), I got this failure in the 7th iteration:

```
--- FAIL: TestIntegration (0.19s)
    --- FAIL: TestIntegration/sync/pull_rebase_interactive_conflict (0.65s)
        go_test.go:46:
            	Error Trace:	/Users/stk/Stk/Dev/Builds/lazygit/go_test.go:46
            	Error:      	Received unexpected error:
            	            	fatal error: concurrent map read and map write
```

<details>
  <summary>Full log</summary>

  ```
--- FAIL: TestIntegration (0.19s)
    --- FAIL: TestIntegration/sync/pull_rebase_interactive_conflict (0.65s)
        go_test.go:46:
            	Error Trace:	/Users/stk/Stk/Dev/Builds/lazygit/go_test.go:46
            	Error:      	Received unexpected error:
            	            	fatal error: concurrent map read and map write

            	            	goroutine 99 [running]:
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_config.(*CachedGitConfig).Get(0x1400000c1e0, {0x10270cc88, 0x19})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_config/cached_git_config.go:38 +0x44
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*ConfigCommands).GetShowUntrackedFiles(0x14000487c78?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/config.go:81 +0x34
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*FileLoader).GetStatusFiles(0x14000404510, {0x0?})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/file_loader.go:38 +0x3c
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshStateFiles(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:443 +0x26c
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshFilesAndSubmodules(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:390 +0x120
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func1.5()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:116 +0x20
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func1.1.1(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:91 +0x24
            	            	github.com/jesseduffield/gocui.(*Gui).onWorkerAux(0x0?, 0x0?, 0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:741 +0x60
            	            	github.com/jesseduffield/gocui.(*Gui).OnWorker.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:728 +0x2c
            	            	created by github.com/jesseduffield/gocui.(*Gui).OnWorker
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:727 +0x9c

            	            	goroutine 1 [runnable]:
            	            	github.com/jesseduffield/gocui.tcellSetCell(0x8?, 0x31?, 0x20?, 0x0?, 0x0?, 0x4?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/tcell_driver.go:101 +0xf4
            	            	github.com/jesseduffield/gocui.(*View).clearRunes(0x14000322780)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/view.go:1124 +0xb0
            	            	github.com/jesseduffield/gocui.(*View).draw(0x14000322780)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/view.go:937 +0xc8
            	            	github.com/jesseduffield/gocui.(*Gui).draw(0x14000438000, 0x14000322780)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:1251 +0x178
            	            	github.com/jesseduffield/gocui.(*Gui).flush(0x14000438000)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:1203 +0x110
            	            	github.com/jesseduffield/gocui.(*Gui).processEvent(0x14000438000)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:827 +0x1ec
            	            	github.com/jesseduffield/gocui.(*Gui).MainLoop(0x14000438000)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:800 +0xa0
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*Gui).Run(0x14000434000, {{0x0, 0x0}, {0x0, 0x0}, {0x1029fa6e0, 0x140002defa0}})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/gui.go:630 +0x528
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*Gui).RunAndHandleError.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/gui.go:636 +0x40
            	            	github.com/jesseduffield/lazygit/pkg/utils.SafeWithError(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:119 +0x60
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*Gui).RunAndHandleError(0x14000434000, {{0x0, 0x0}, {0x0, 0x0}, {0x1029fa6e0, 0x140002defa0}})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/gui.go:635 +0xbc
            	            	github.com/jesseduffield/lazygit/pkg/app.(*App).Run(...)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/app/app.go:258
            	            	github.com/jesseduffield/lazygit/pkg/app.Run({0x102a03750?, 0x14000315ae0?}, 0x140002f6c00, {{0x0, 0x0}, {0x0, 0x0}, {0x1029fa6e0, 0x140002defa0}})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/app/app.go:48 +0x8c
            	            	github.com/jesseduffield/lazygit/pkg/app.Start(0x140001d3f28, {0x1029fa6e0, 0x140002defa0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/app/entry_point.go:151 +0x9b8
            	            	main.main()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/clients/injector/main.go:32 +0x50

            	            	goroutine 25 [select]:
            	            	github.com/jesseduffield/gocui.(*Gui).pollEvent(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/tcell_driver.go:241 +0x88
            	            	github.com/jesseduffield/gocui.(*Gui).MainLoop.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:790 +0x30
            	            	created by github.com/jesseduffield/gocui.(*Gui).MainLoop
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:784 +0x60

            	            	goroutine 27 [chan receive]:
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*GuiDriver).PressKey(0x140008ee580, {0x1026e090c?, 0x102636800?})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/gui_driver.go:46 +0x118
            	            	github.com/jesseduffield/lazygit/pkg/integration/components.(*TestDriver).press(0x140008dcd00, {0x1026e090c, 0x1})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/components/test_driver.go:35 +0x5c
            	            	github.com/jesseduffield/lazygit/pkg/integration/components.(*ViewDriver).Press(0x140008a6ec0, {0x1026e090c, 0x1})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/components/view_driver.go:367 +0x3c
            	            	github.com/jesseduffield/lazygit/pkg/integration/tests/sync.glob..func33(_, {{{0x1026e090d, 0x1}, {0x1026e32bf, 0x5}, {0x1026e332d, 0x5}, {0x1028592e8, 0x1}, {0x1026e3346, ...}, ...}, ...})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/tests/sync/pull_rebase_interactive_conflict.go:43 +0x23c
            	            	github.com/jesseduffield/lazygit/pkg/integration/components.(*IntegrationTest).Run(0x140002defa0, {0x102a037c0?, 0x140008ee580})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/components/test.go:167 +0x2b0
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*Gui).handleTestMode.func2()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/test_mode.go:35 +0xb0
            	            	created by github.com/jesseduffield/lazygit/pkg/gui.(*Gui).handleTestMode
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/test_mode.go:32 +0x208

            	            	goroutine 28 [sleep]:
            	            	time.Sleep(0x9502f9000)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/time.go:195 +0x118
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*Gui).handleTestMode.func3()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/test_mode.go:49 +0x28
            	            	github.com/jesseduffield/lazygit/pkg/utils.Safe.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:108 +0x24
            	            	github.com/jesseduffield/lazygit/pkg/utils.SafeWithError(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:119 +0x60
            	            	github.com/jesseduffield/lazygit/pkg/utils.Safe(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:108 +0x34
            	            	created by github.com/jesseduffield/lazygit/pkg/gui.(*Gui).handleTestMode
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/test_mode.go:48 +0x230

            	            	goroutine 98 [semacquire]:
            	            	sync.runtime_Semacquire(0x140001024e0?)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/sema.go:62 +0x2c
            	            	sync.(*WaitGroup).Wait(0x140002d43c0)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/sync/waitgroup.go:116 +0x78
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshCommits(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:249 +0x174
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func1.1.1(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:91 +0x24
            	            	github.com/jesseduffield/gocui.(*Gui).onWorkerAux(0x0?, 0x1020925b4?, 0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:741 +0x60
            	            	github.com/jesseduffield/gocui.(*Gui).OnWorker.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:728 +0x2c
            	            	created by github.com/jesseduffield/gocui.(*Gui).OnWorker
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:727 +0x9c

            	            	goroutine 145 [runnable]:
            	            	github.com/jesseduffield/gocui.(*Gui).StartTicking.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:1497 +0x160
            	            	created by github.com/jesseduffield/gocui.(*Gui).StartTicking
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:1490 +0x94

            	            	goroutine 73 [runnable]:
            	            	os/exec.(*Cmd).Start.func3()
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:716
            	            	runtime.goexit()
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/asm_arm64.s:1172 +0x4
            	            	created by os/exec.(*Cmd).Start
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:716 +0x7f4

            	            	goroutine 177 [runnable]:
            	            	github.com/sasha-s/go-deadlock.lock.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/sasha-s/go-deadlock/deadlock.go:182
            	            	created by github.com/sasha-s/go-deadlock.lock
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/sasha-s/go-deadlock/deadlock.go:182 +0x170

            	            	goroutine 162 [runnable]:
            	            	bufio.(*Scanner).Scan(0x140003fb828)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/bufio/scan.go:204 +0x3f0
            	            	github.com/fsmiamoto/git-todo-parser/todo.Parse({0x1029f73e0?, 0x140003ac750?}, 0x10?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/fsmiamoto/git-todo-parser/todo/parse.go:25 +0x94
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*CommitLoader).getInteractiveRebasingCommits(0x140000842a0)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/commit_loader.go:310 +0x118
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*CommitLoader).getRebasingCommits(0x14000484ae8?, 0x10210c090?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/commit_loader.go:244 +0x5c
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*CommitLoader).getHydratedRebasingCommits(0x140000842a0, 0x1032e4a68?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/commit_loader.go:192 +0x28
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*CommitLoader).MergeRebasingCommits(0x140000842a0, {0x102fff1c8, 0x0, 0x0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/commit_loader.go:136 +0x120
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*CommitLoader).GetCommits(0x140000842a0, {0x1, {0x0, 0x0}, 0x1, {0x1026e182d, 0x4}, 0x0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/commit_loader.go:80 +0x7c
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshCommitsWithLimit(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:256 +0x274
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshCommits.func2()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:230 +0x2c
            	            	github.com/jesseduffield/lazygit/pkg/utils.Safe.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:108 +0x24
            	            	github.com/jesseduffield/lazygit/pkg/utils.SafeWithError(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:119 +0x60
            	            	github.com/jesseduffield/lazygit/pkg/utils.Safe(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:108 +0x34
            	            	created by github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshCommits
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:229 +0x16c

            	            	goroutine 161 [runnable]:
            	            	internal/poll.runtime_pollWait(0x12a6b7a68, 0x72)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/netpoll.go:306 +0xa0
            	            	internal/poll.(*pollDesc).wait(0x14000886f00?, 0x14000aa4000?, 0x1)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/internal/poll/fd_poll_runtime.go:84 +0x28
            	            	internal/poll.(*pollDesc).waitRead(...)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/internal/poll/fd_poll_runtime.go:89
            	            	internal/poll.(*FD).Read(0x14000886f00, {0x14000aa4000, 0x1000, 0x1000})
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/internal/poll/fd_unix.go:167 +0x200
            	            	os.(*File).read(...)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/file_posix.go:31
            	            	os.(*File).Read(0x14000884208, {0x14000aa4000?, 0x10?, 0x14000882c60?})
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/file.go:118 +0x5c
            	            	bufio.(*Scanner).Scan(0x14000485c20)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/bufio/scan.go:214 +0x948
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*cmdObjRunner).RunAndProcessLines(0x20?, {0x102a0af40, 0x140008faf00}, 0x1400088b200)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj_runner.go:160 +0x250
            	            	github.com/jesseduffield/lazygit/pkg/commands.(*gitCmdObjRunner).RunAndProcessLines(0x20?, {0x102a0af40?, 0x140008faf00?}, 0x140008faf00?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_cmd_obj_runner.go:29 +0x30
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*CmdObj).RunAndProcessLines(0x140008eef80?, 0x14000a96af0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj.go:195 +0x3c
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*ReflogCommitLoader).GetReflogCommits(0x1400000c288, 0x140004181e0, {0x0, 0x0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/reflog_commit_loader.go:39 +0x5a4
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshReflogCommits.func1(0x14000000358, {0x0, 0x0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:500 +0x60
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshReflogCommits(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:513 +0x8c
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshReflogCommitsConsideringStartup(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:211 +0x60
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshCommits.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:223 +0x2c
            	            	github.com/jesseduffield/lazygit/pkg/utils.Safe.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:108 +0x24
            	            	github.com/jesseduffield/lazygit/pkg/utils.SafeWithError(0x1400044377e?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:119 +0x60
            	            	github.com/jesseduffield/lazygit/pkg/utils.Safe(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/utils/utils.go:108 +0x34
            	            	created by github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshCommits
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:222 +0xd0

            	            	goroutine 100 [runnable]:
            	            	github.com/jesseduffield/gocui.(*Gui).OnWorker.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:727
            	            	created by github.com/jesseduffield/gocui.(*Gui).OnWorker
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:727 +0x9c

            	            	goroutine 101 [syscall]:
            	            	syscall.syscall6(0x1032e6b38?, 0x17?, 0x5?, 0x14000809800?, 0x140008098e8?, 0x900001010185b4?, 0x14000a96c40?)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/sys_darwin.go:45 +0x68
            	            	syscall.wait4(0x14000809898?, 0x102106698?, 0x90?, 0x1029c96a0?)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/syscall/zsyscall_darwin_arm64.go:43 +0x4c
            	            	syscall.Wait4(0x140004869b8?, 0x140008098d4, 0x14000392258?, 0x14000884240?)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/syscall/syscall_bsd.go:144 +0x28
            	            	os.(*Process).wait(0x140001361e0)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec_unix.go:43 +0x80
            	            	os.(*Process).Wait(...)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec.go:132
            	            	os/exec.(*Cmd).Wait(0x140008d8840)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:890 +0x38
            	            	os/exec.(*Cmd).Run(0x0?)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:590 +0x38
            	            	os/exec.(*Cmd).CombinedOutput(0x140008d8840)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:1005 +0xb8
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*cmdObjRunner).RunWithOutputAux(0x14000410120, {0x102a0af40, 0x140008faff0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj_runner.go:105 +0x148
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*cmdObjRunner).RunWithOutput(0x90?, {0x102a0af40, 0x140008faff0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj_runner.go:70 +0x1ac
            	            	github.com/jesseduffield/lazygit/pkg/commands.(*gitCmdObjRunner).RunWithOutput(0x1?, {0x102a0af40?, 0x140008faff0?})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_cmd_obj_runner.go:21 +0x30
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*CmdObj).RunWithOutput(0x140008eefe0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj.go:187 +0x38
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*TagLoader).GetTags(0x1400000c2b8)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/tag_loader.go:32 +0x20c
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshTags(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:324 +0x3c
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func1.7()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:124 +0x20
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func1.1.1(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:91 +0x24
            	            	github.com/jesseduffield/gocui.(*Gui).onWorkerAux(0x0?, 0x0?, 0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:741 +0x60
            	            	github.com/jesseduffield/gocui.(*Gui).OnWorker.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:728 +0x2c
            	            	created by github.com/jesseduffield/gocui.(*Gui).OnWorker
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:727 +0x9c

            	            	goroutine 102 [runnable]:
            	            	sync.runtime_SemacquireRWMutexR(0x10?, 0x60?, 0x14000805701?)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/sema.go:82 +0x28
            	            	sync.(*RWMutex).RLock(...)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/sync/rwmutex.go:71
            	            	os.Pipe()
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/pipe_unix.go:17 +0x7c
            	            	os/exec.(*Cmd).writerDescriptor(0x1400048e2c0, {0x1029f7400, 0x14000b0fe00})
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:552 +0x5c
            	            	os/exec.(*Cmd).childStdout(...)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:524
            	            	os/exec.(*Cmd).Start(0x1400048e2c0)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:676 +0x2b0
            	            	os/exec.(*Cmd).Run(0x0?)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:587 +0x20
            	            	os/exec.(*Cmd).CombinedOutput(0x1400048e2c0)
            	            		/opt/homebrew/Cellar/go/1.20.5/libexec/src/os/exec/exec.go:1005 +0xb8
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*cmdObjRunner).RunWithOutputAux(0x14000410120, {0x102a0af40, 0x1400049cb90})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj_runner.go:105 +0x148
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*cmdObjRunner).RunWithOutput(0x90?, {0x102a0af40, 0x1400049cb90})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj_runner.go:70 +0x1ac
            	            	github.com/jesseduffield/lazygit/pkg/commands.(*gitCmdObjRunner).RunWithOutput(0x1?, {0x102a0af40?, 0x1400049cb90?})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_cmd_obj_runner.go:21 +0x30
            	            	github.com/jesseduffield/lazygit/pkg/commands/oscommands.(*CmdObj).RunWithOutput(0x14000b0d2f0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/oscommands/cmd_obj.go:187 +0x38
            	            	github.com/jesseduffield/lazygit/pkg/commands/git_commands.(*RemoteLoader).GetRemotes(0x14000412320)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/commands/git_commands/remote_loader.go:35 +0x1dc
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshRemotes(0x14000031600)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:531 +0x74
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func1.8()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:128 +0x20
            	            	github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func1.1.1(0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:91 +0x24
            	            	github.com/jesseduffield/gocui.(*Gui).onWorkerAux(0x0?, 0x0?, 0x0?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:741 +0x60
            	            	github.com/jesseduffield/gocui.(*Gui).OnWorker.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:728 +0x2c
            	            	created by github.com/jesseduffield/gocui.(*Gui).OnWorker
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:727 +0x9c

            	            	goroutine 107 [runnable]:
            	            	github.com/sasha-s/go-deadlock.lock.func1()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/sasha-s/go-deadlock/deadlock.go:182
            	            	created by github.com/sasha-s/go-deadlock.lock
            	            		/Users/stk/Stk/Dev/Builds/lazygit/vendor/github.com/sasha-s/go-deadlock/deadlock.go:182 +0x170
            	Test:       	TestIntegration/sync/pull_rebase_interactive_conflict
  ```
</details>

stefanhaller commented:
I pushed a trivial fixup for `CI.yml`.

jesseduffield commented:
Thanks for that commit, I've merged it into https://github.com/jesseduffield/lazygit/pull/2765/commits/c7a3b69eb92dd5b543769cafcb2992da1ee8ed4d

Updated the doc in https://github.com/jesseduffield/lazygit/pull/2765/commits/94800a5d53231e5c2dd1cd3dfc7980f030a47c2c

Fixed the race condition (I'm pretty sure) in https://github.com/jesseduffield/lazygit/commit/24dc870c4ebbe21d31499a1231ddabaa7699096d

stefanhaller commented on docs/dev/Busy.md:
````diff
@@ -25,27 +25,27 @@ First, it's important to distinguish three different types of goroutines:
 ```
 
-The crucial thing here is that we increment the busy count _before_ spawning the goroutine, because it means that our counter never goes to zero while there's still work being done. If we incremented the busy count within the goroutine, the current function could exit and decrement the counter to zero before the goroutine starts.
+The crucial thing here is that we create the task _before_ spawning the goroutine, because it means that we'll have at least one busy task in the map until the completion of the goroutine. created the task within the goroutine, the current function could exit and Lazygit would be considered idle before the goroutine starts, leading to our integration test prematurely progressing.
````
There seem to be words missing here. "_If we_ created the task within the goroutine", maybe?

jesseduffield replied:
Good pickup

stefanhaller commented:
I still got a failure in `initial_open.go`, after like 15 iterations.

<details>
  <summary>Test log</summary>

```
--- FAIL: TestIntegration (0.18s)
    runner.go:94: Skipping test patch_building/move_to_earlier_commit_no_keep_empty for git version 2.41.0
    runner.go:89: Skipping test patch_building/copy_patch_to_clipboard
    --- FAIL: TestIntegration/misc/initial_open (0.34s)
        go_test.go:46:
            	Error Trace:	/Users/stk/Stk/Dev/Builds/lazygit/go_test.go:46
            	Error:      	Received unexpected error:
            	            	files view: Unexpected view focused. Expected files, got confirmation
            	            	Final Lazygit state:
            	            	┌─Status─────────────────────────┐┌─Diff───────────────────────────────────────────────────────────┐
            	            	│repo → master                   ││No changed files                                                │
            	            	└────────────────────────────────┘│                                                                │
            	            	┌─Files - Submodules─────────────┐│                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	└─────────────────────────0 of 0─┘│                                                                │
            	            	┌─Local branches - Remotes - Tag─┐│                                                                │
            	            	│  * master                      ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│         ┌───────────────────────────────────────────────────────────────────────────────┐        │
            	            	│         │                                                                               │        │
            	            	│         │Thanks for using lazygit! Seriously you rock. Three things to share with you:  │        │
            	            	│         │                                                                               │        │
            	            	│         │ 1) If you want to learn about lazygit's features, watch this vid:             │        │
            	            	│         │      https://youtu.be/CPLdltN7wgE                                             │        │
            	            	│         │                                                                               │        │
            	            	│         │ 2) Be sure to read the latest release notes at:                               │        │
            	            	│         │      https://github.com/jesseduffield/lazygit/releases                        │        │
            	            	│         │                                                                               │        │
            	            	│         │ 3) If you're using git, that makes you a programmer! With your help we can mak│        │
            	            	│         │e                                                                              │        │
            	            	│         │    lazygit better, so consider becoming a contributor and joining the fun at  │        │
            	            	│         │      https://github.com/jesseduffield/lazygit                                 │        │
            	            	│         │    You can also sponsor me and tell me what to work on by clicking the donate │        │
            	            	│         │    button at the bottom right.                                                │        │
            	            	│         │    Or even just star the repo to share the love!                              │        │
            	            	│         │                                                                               │        │
            	            	│         └───────────────────────────────────────────────────────────────────────────────┘        │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	└─────────────────────────1 of 1─┘│                                                                │
            	            	┌─Commits - Reflog───────────────┐│                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                ││                                                                │
            	            	│                                │└────────────────────────────────────────────────────────────────┘
            	            	│                                │┌─Command log────────────────────────────────────────────────────┐
            	            	│                                ││You can hide/focus this panel by pressing '@'                   │
            	            	│                                ││                                                                │
            	            	│                                ││Random tip: The stash is a good place to save snippets of code t│
            	            	│                                ││hat you always find yourself adding when debugging.             │
            	            	│                                ││                                                                │
            	            	└─────────────────────────0 of 0─┘│                                                                │
            	            	┌─Stash──────────────────────────┐│                                                                │
            	            	│                                ││                                                                │
            	            	└─────────────────────────0 of 0─┘└────────────────────────────────────────────────────────────────┘
            	            	 <pgup>/<pgdown>: Scroll, <esc>: Cancel, q: Quit, ?: Keybindings, 1-Donate Ask Question unversioned

            	            	Upon failure, focused view was 'files'.
            	            	Log:

            	            	panic: Test failed

            	            	goroutine 10 [running]:
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*GuiDriver).Fail(0x14000390db0, {0x14000b09db0, 0x45})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/gui_driver.go:82 +0x1e8
            	            	github.com/jesseduffield/lazygit/pkg/integration/components.(*assertionHelper).fail(...)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/components/assertion_helper.go:28
            	            	github.com/jesseduffield/lazygit/pkg/integration/components.(*assertionHelper).assertWithRetries(0x14000390dc0, 0x14000097370?)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/components/assertion_helper.go:23 +0x44
            	            	github.com/jesseduffield/lazygit/pkg/integration/components.(*ViewDriver).IsFocused(0x14000396d00)
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/components/view_driver.go:355 +0x44
            	            	github.com/jesseduffield/lazygit/pkg/integration/tests/misc.glob..func6(_, {{{0x10075cbcd, 0x1}, {0x10075f57f, 0x5}, {0x10075f5ed, 0x5}, {0x1008d55e8, 0x1}, {0x10075f606, ...}, ...}, ...})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/tests/misc/initial_open.go:22 +0x114
            	            	github.com/jesseduffield/lazygit/pkg/integration/components.(*IntegrationTest).Run(0x140002daa00, {0x100a7f7c0?, 0x14000390db0})
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/integration/components/test.go:167 +0x2b0
            	            	github.com/jesseduffield/lazygit/pkg/gui.(*Gui).handleTestMode.func2()
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/test_mode.go:35 +0xb0
            	            	created by github.com/jesseduffield/lazygit/pkg/gui.(*Gui).handleTestMode
            	            		/Users/stk/Stk/Dev/Builds/lazygit/pkg/gui/test_mode.go:32 +0x208
            	Test:       	TestIntegration/misc/initial_open
  ```
</details>

jesseduffield commented:
@stefanhaller damn. I'll take a look

jesseduffield commented:
Yep, I've got the following tests intermittently failing (not very often)
```
patch_building/start_new_patch.go
patch_building/move_to_index_with_conflict
patch_building/move_to_index_partial
file/discard_staged_changes
file/discard_all_dir_changes
```

The discard changes ones are from index.lock issues, and the others seem to be from a main panel not refreshing before the test runs an assertion. I've spent some time on both of these issues have can't work out what the issue is.

With the index.lock issue I'm especially confused: we're not running these commands concurrently and I doubt the issue is that some other git process died midway and left behind the lock file because if that was a common thing we should see it in other tests

stefanhaller commented:
> and I doubt the issue is that some other git process died midway and left behind the lock file because if that was a common thing we should see it in other tests

Also, the file would then still exist after the test finished. It didn't in my case.

jesseduffield commented:
Ah that's good to know.

I've tested out adding a global mutex to `pkg/commands/oscommands/cmd_obj_runner.go` so that no two commands can run at the same time and I still get the index.lock error.

stefanhaller commented:
That's very weird.

I don't have any great ideas how to diagnose this. Except maybe add some debugging code that dumps the output of `pstree` whenever a command fails, so that we can get an idea what git process it is that's running concurrently.

jesseduffield commented:
Wish I was in the habit of writing better commit messages back then! The PR description gives a clue: we were getting a useless error appear on some commands (I think ones that stream output perhaps) and the lock retry logic was behind it. Though I don't recall the specifics. I'm happy to re-instate it and see if the other issue pops up again.

jesseduffield commented:
Ah, just tested with that and found the issue again: `exec: Stdout already set`. Digging deeper

jesseduffield commented:
Okay so the issue there was that we need to clone the cmd in order to re-use it (something I didn't realise or wasn't bothered to fix when I originally removed the retry logic).

That's fixed the issue on my end.

jesseduffield commented:
I was able to get 24 runs of the test suite before hitting another failure (`stash/stash_staged`). I think we've got less flakyness with this PR than we have on master so I'm happy to merge and investigate more flakyness later.

Though I'm conscious that the PR has some sweeping changes so I'd like to take a look at your PRs before merging this one so that you're not swamped with merge conflicts (apologies for not getting around to them yet)

stefanhaller commented:
Nice. I just ran this for over 100 iterations without failures. I agree that this looks less flaky now than what we had on master.

I'm happy for this to get merged. Don't worry about my branches so much, I can rebase them onto this once it's merged. (And then I'd appreciate reviews at some point. 😄)
